
syntax = "proto3";

package NetworkIO;

import "Types.proto";
import "Effects.proto";
import "GameState.proto";

message CardInfo {
    GameState.CardMetaData metadata = 1;
    int32 Count = 2;
}

message RandomAgent {

}

message AIAgent {
    
}

// the game doesn't talk directly to clients, it instead talks to the server, which talks to the clients
message PlayerAgent {
}

message PlayerInfo {
    string Username = 1;
    string UID = 2;
    repeated CardInfo Deck = 3;
    Types.Vec2i StartingPosition = 4;
    oneof AgentInfo {
        AIAgent AIAgent = 5;
        RandomAgent RandomAgent = 6;
        PlayerAgent PlayerAgent = 7;
    }
}

message GameInfo {
    repeated PlayerInfo Players = 1;
    Types.Vec2i MapSize = 2;
}